File: src/clu/sources/macos_name.py
import logging

from clu import panic
from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.sources import PARSE_FAIL_MSG

log = logging.getLogger(__name__)


class MacOSName(Source):
    def provides(self, provides: Provides) -> None:
        provides["os.code_name"] = self

    def requires(self, requires: Requires) -> None:
        requires.facts.append("os.version")

    def parse(self, facts: Facts) -> None:
        version = facts["os.version"]
        if not version:
            panic("parse_macos_name: os.version is not set or empty")

        major_ver = version.split(".")[0]
        log.debug(f"{major_ver=}")

        if major_ver == "26":
            code_name = "Tahoe"
        elif major_ver == "15":
            code_name = "Sequoia"
        elif major_ver == "14":
            code_name = "Sonoma"
        elif major_ver == "13":
            code_name = "Ventura"
        elif major_ver == "12":
            code_name = "Monterey"
        elif major_ver == "11":
            code_name = "Big Sur"
        else:
            # Note that for older than 11, the logic of the code name changes
            # and thats WAY out of support for us
            code_name = PARSE_FAIL_MSG
        facts["os.code_name"] = code_name

File: pkg/facts/sources/macos_name.go
package sources

import (
	"strings"

	"github.com/huntermatthews/clu/pkg/facts/types"
)

// MacOSName derives the macOS marketing code name from the major version.
type MacOSName struct{}

var codeNameFact = types.Fact{
	Name: "os.code_name",
	Tier: types.TierOne,
}

func (m *MacOSName) Provides(p types.Provides) { p[codeNameFact.Name] = m }

func (m *MacOSName) Requires(r *types.Requires) { r.Facts = append(r.Facts, "os.version") }

func (m *MacOSName) Parse(f *types.FactDB) {
	ver, ok := f.Get("os.version")
	if !ok || strings.TrimSpace(ver) == "" {
		codeNameFact.Value = types.ParseFailMsg
		f.AddFact(codeNameFact)
		return
	}

	major := strings.Split(ver, ".")[0]
	codeNameFact.Value = codeNameFromMajor(major)
	f.AddFact(codeNameFact)
}

func codeNameFromMajor(major string) string {
	switch major {
	case "26":
		return "Tahoe"
	case "15":
		return "Sequoia"
	case "14":
		return "Sonoma"
	case "13":
		return "Ventura"
	case "12":
		return "Monterey"
	case "11":
		return "Big Sur" // big sur is OLD enough...
	default:
		return types.ParseFailMsg
	}
}
